docs/blog/2021-12-01-csp.md

```diff
@@ -0,0 +1,153 @@
+---
+title: Don't trust your Content Security Policy
+description: Why the CSP most sites have is useless.
+slug: csp
+authors:
+- name: Forrest Allison
+  title: Developer at LunaSec
+  url: https://github.com/factoidforrest
+  image_url: https://github.com/factoidforrest.png
+  tags: [ci, github-actions, docker.sock]
+---
+<!--
+  ~ Copyright by LunaSec (owned by Refinery Labs, Inc)
+  ~
+  ~ Licensed under the Creative Commons Attribution-ShareAlike 4.0 International
+  ~ (the "License"); you may not use this file except in compliance with the
+  ~ License. You may obtain a copy of the License at
+  ~
+  ~ https://creativecommons.org/licenses/by-sa/4.0/legalcode
+  ~
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  ~
+-->
+
+I'm a normal, non-security focused developer, but I work for a security company.  I wasn't surprised to learn that our 
+services use a good, well-thought-out CSP to protect our users. 
+
+What did come as a *big* surprise to me is that most companies, even huge, well thought of firms, have CSPs that don't
+protect from *anything*, or don't even use them at all. For the most part, it's not their fault and there's **not much they can do
+about it.**  More on that below.
+
+### How CSPs work
+
+The Content Security Policy (CSP) is a response header or `<meta>` tag set by the server when a webpage first loads.  It controls 
+what the page is allowed to load and, most importantly, from _where_.  The idea is to act as a last-line-of-defense against
+ loading something malicious onto your page.  
+
+If you've never seen one before, they're actually pretty simple.
+
+```text title="Example small CSP"
+Content-Security-Policy: 
+    default-src 'self'; 
+    script-src 'self' www.google.com;
+```
+
+<sub>For a more complete overview of CSP directives and values, this is a <a href="https://www.arridae.com/blogs/bypass-csp.php">great article</a>.</sub>  
+
+
+
+The CSP above will make it so that your page can only load any content from your domain, with the exception that scripts can also load from
+Google. It's pretty common to include a few additional domains (like Google) in here that you need to load things from,
+such as analytics or some third-party auth scripts.  
+
+That CSP also means that inline scripts like `<script> alert('Hi I'm a script') <script>` won't be executed, which is great for security.
+
+So, if we trust our server and Google's server not to do something malicious, we would expect to be safe.  
+
+Unfortunately, this CSP is still totally insecure, because:
+
+## CSP is swiss cheese
+Studies have found that the vast majority of sites using CSPs are still vulnerable to script injection.  Here are some of the most common attacks.
+### JSONP Bypass
+JSONP is a hacky way to load cross domain javascript in the browser from the mid 2000's, that has since been replaced by CORS.  More of a pattern than a protocol, 
```

Comment:
Extraneous comma here (the second half after the comma is a continuation and not a dependent clause).

Also, I substituted "replaced" with "superceded" because I believe that's clearer. They both exist in parallel, but JSONP is legacy.

Finally, I replaced `javascript` with `JavaScript` because that feels a little better. I'd also accept `JS` too.

https://www.scribbr.com/language-rules/commas/

Options:

**Rewrite second half to be a dependent clause**:

> JSONP is a hacky way to load cross domain JavaScript in the browser from the mid 2000's, and it has since been superceded by CORS.

**Remove comma to make it one long sentence**:

> JSONP is a hacky way to load cross domain JavaScript in the browser from the mid 2000's that has since been superceded by CORS.

**Make the middle a dependent clause**:
> JSONP is a hacky way to load cross domain JavaScript in the browser, added in the mid 2000's, that has since been superceded by CORS.

That's my thought process on this. Small tweaks like this will make the text flow a lot more nicely, imo.